server/src/main/java/org/opensearch/search/pipeline/Pipeline.java

```diff
@@ -0,0 +1,171 @@
+    private static <T extends Processor> List<T> readProcessors(
+        Class<T> processorType,
+        Map<String, Processor.Factory> processorFactories,
+        List<Map<String, Object>> requestProcessorConfigs
+    ) throws Exception {
+        List<T> processors = new ArrayList<>();
+        if (requestProcessorConfigs == null) {
+            return processors;
+        }
+        for (Map<String, Object> processorConfigWithKey : requestProcessorConfigs) {
+            for (Map.Entry<String, Object> entry : processorConfigWithKey.entrySet()) {
+                String type = entry.getKey();
+                if (!processorFactories.containsKey(type)) {
+                    throw new IllegalArgumentException("Invalid processor type " + type);
+                }
+                Map<String, Object> config = (Map<String, Object>) entry.getValue();
+                String tag = ConfigurationUtils.readOptionalStringProperty(null, null, config, TAG_KEY);
+                String description = ConfigurationUtils.readOptionalStringProperty(null, tag, config, DESCRIPTION_KEY);
+                Processor processor = processorFactories.get(type).create(processorFactories, tag, description, config);
+                if (processorType.isInstance(processor)) {
+                    processors.add((T) processor);
+                } else {
+                    throw new IllegalArgumentException("Processor type " + type + " is not a " + processorType.getSimpleName());
+                }
+            }
+        }
+        return processors;
+    }
+
+    List<Processor> flattenAllProcessors() {
+        List<Processor> allProcessors = new ArrayList<>(searchRequestProcessors.size() + searchResponseProcessors.size());
+        allProcessors.addAll(searchRequestProcessors);
+        allProcessors.addAll(searchResponseProcessors);
+        return allProcessors;
+    }
+
+    String getId() {
+        return id;
+    }
+
+    String getDescription() {
+        return description;
+    }
+
+    Integer getVersion() {
+        return version;
+    }
+
+    List<SearchRequestProcessor> getSearchRequestProcessors() {
+        return searchRequestProcessors;
+    }
+
+    List<SearchResponseProcessor> getSearchResponseProcessors() {
+        return searchResponseProcessors;
+    }
+
+    SearchRequest transformRequest(SearchRequest originalRequest) throws SearchPipelineProcessingException {
+        try {
+            // Save the original request by deep cloning the existing request.
+            BytesStreamOutput bytesStreamOutput = new BytesStreamOutput();
+            originalRequest.writeTo(bytesStreamOutput);
+            SearchRequest request = new SearchRequest(bytesStreamOutput.bytes().streamInput());
+            for (SearchRequestProcessor searchRequestProcessor : searchRequestProcessors) {
+                request = searchRequestProcessor.processRequest(request);
+            }
+            return request;
```

Comment:
one optimization I would recommend here is we should not start cloning the object, if there are no SearchRequestProcessor present in the pipeline.

Reply:
Good idea! Added in the next commit